lib/common/tests/rules/pcmk__evaluate_op_expression_test.c

```diff
@@ -0,0 +1,180 @@
+    // These are the only members used to evaluate operation expressions
+    .op_name = PCMK_ACTION_MONITOR,
+    .op_interval_ms = 10000,
+};
+
+/*!
+ * \internal
+ * \brief Run one test, comparing return value
+ *
+ * \param[in] xml_string    Operation expression XML as string
+ * \param[in] reference_rc  Assert that evaluation result equals this
+ */
+static void
+assert_op_expression(const char *xml_string, int reference_rc)
+{
+    xmlNode *xml = pcmk__xml_parse(xml_string);
+
+    assert_int_equal(pcmk__evaluate_op_expression(xml, &rule_input),
+                     reference_rc);
+    free_xml(xml);
+}
+
+
+/*
+ * Invalid arguments
+ */
+
+#define EXPR_NO_MATCH                                       \
+        "<" PCMK_XE_OP_EXPRESSION " " PCMK_XA_ID "='e' "    \
+        PCMK_XA_NAME "='" PCMK_ACTION_START "' "            \
+        PCMK_XA_INTERVAL "='0' />"
+
+static void
+null_invalid(void **state)
+{
+    xmlNode *xml = NULL;
+
+    assert_int_equal(pcmk__evaluate_op_expression(NULL, NULL), EINVAL);
+
+    xml = pcmk__xml_parse(EXPR_NO_MATCH);
+    assert_int_equal(pcmk__evaluate_op_expression(xml, NULL), EINVAL);
+    free_xml(xml);
+
+    assert_op_expression(NULL, EINVAL);
+}
+
+
+/*
+ * Test PCMK_XA_ID
+ */
+
+#define EXPR_ID_MISSING                                 \
+        "<" PCMK_XE_OP_EXPRESSION " "                   \
+        PCMK_XA_NAME "='" PCMK_ACTION_MONITOR "' "      \
+        PCMK_XA_INTERVAL "='10s' />"
+
+static void
+id_missing(void **state)
+{
+    // Currently acceptable
+    assert_op_expression(EXPR_ID_MISSING, pcmk_rc_ok);
+}
+
+
+/*
+ * Test PCMK_XA_NAME
+ */
+
+#define EXPR_NAME_MISSING                                   \
+        "<" PCMK_XE_OP_EXPRESSION " " PCMK_XA_ID "='e' "    \
+        PCMK_XA_INTERVAL "='10s' />"
+
+static void
+name_missing(void **state)
+{
+    assert_op_expression(EXPR_NAME_MISSING, pcmk_rc_unpack_error);
+}
+
+#define EXPR_ALL_MATCH                                      \
+        "<" PCMK_XE_OP_EXPRESSION " " PCMK_XA_ID "='e' "    \
+        PCMK_XA_NAME "='" PCMK_ACTION_MONITOR "' "          \
+        PCMK_XA_INTERVAL "='10s' />"
+
+static void
+reference_name_missing(void **state)
+{
+    rule_input.op_name = NULL;
+    assert_op_expression(EXPR_ALL_MATCH, pcmk_rc_op_unsatisfied);
+    rule_input.op_name = PCMK_ACTION_MONITOR;
+}
+
+#define EXPR_FAIL_NAME                                      \
+        "<" PCMK_XE_OP_EXPRESSION " " PCMK_XA_ID "='e' "    \
+        PCMK_XA_NAME "='" PCMK_ACTION_START "' "            \
+        PCMK_XA_INTERVAL "='10s' />"
+
+static void
+fail_name(void **state)
+{
+    assert_op_expression(EXPR_FAIL_NAME, pcmk_rc_op_unsatisfied);
+}
+
+
+/*
+ * Test PCMK_XA_INTERVAL
+ */
+
+#define EXPR_INVALID_INTERVAL                               \
+        "<" PCMK_XE_OP_EXPRESSION " " PCMK_XA_ID "='e' "    \
+        PCMK_XA_NAME "='" PCMK_ACTION_MONITOR "' "          \
+        PCMK_XA_INTERVAL "='not-an-interval' />"
```

Comment:
empty string is also an invalid interval; up to you whether to test it separately or leave it alone

Reply:
thx